### PR TITLE
Fix ansible-test error in community.aws

### PR DESCRIPTION
**Pull request by AlanCoding**
##### SUMMARY
This is a followup from an email on the ansible-devel email list.

This fixes an error I got when running `ansible-test` for `community.aws`.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/schema.py

##### ADDITIONAL INFORMATION
Also in email, the steps for running the sanity tests:

```
cd ~/Documents/test
git clone https://github.com/ansible-collections/community.aws.git
ansible-galaxy collection build
rm -rf ~/.ansible/collections/ansible_collections/community/aws/
ansible-galaxy collection install community-aws-1.0.0.tar.gz
cd ~/.ansible/collections/ansible_collections/community/aws/
ansible-test sanity
```

I tried with Ansible 2.11, 2.10, 2.9, gave up, then realized something was wrong. 

The error I got when running the test looked like this:

```paste below
>>> Standard Error
Traceback (most recent call last):
  File "/Users/alancoding/.virtualenvs/ansible29/lib/python3.7/site-packages/ansible_test/_data/sanity/validate-modules/validate_modules/main.py", line 849, in _validate_docs_schema
    schema(doc)
  File "/Users/alancoding/.virtualenvs/ansible29/lib/python3.7/site-packages/voluptuous/schema_builder.py", line 272, in __call__
    return self._compiled([], data)
  File "/Users/alancoding/.virtualenvs/ansible29/lib/python3.7/site-packages/voluptuous/schema_builder.py", line 594, in validate_dict
    return base_validate(path, iteritems(data), out)
  File "/Users/alancoding/.virtualenvs/ansible29/lib/python3.7/site-packages/voluptuous/schema_builder.py", line 386, in validate_mapping
    cval = cvalue(key_path, value)
  File "/Users/alancoding/.virtualenvs/ansible29/lib/python3.7/site-packages/voluptuous/validators.py", line 205, in _run
    return self._exec(self._compiled, value, path)
  File "/Users/alancoding/.virtualenvs/ansible29/lib/python3.7/site-packages/voluptuous/validators.py", line 285, in _exec
    v = func(path, v)
  File "/Users/alancoding/.virtualenvs/ansible29/lib/python3.7/site-packages/voluptuous/schema_builder.py", line 817, in validate_callable
    return schema(data)
  File "/Users/alancoding/.virtualenvs/ansible29/lib/python3.7/site-packages/ansible_test/_data/sanity/validate-modules/validate_modules/schema.py", line 177, in author
    m = author_line.search(line)
TypeError: expected string or bytes-like object

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/alancoding/.virtualenvs/ansible29/lib/python3.7/site-packages/ansible_test/_data/sanity/validate-modules/validate-modules", line 8, in <module>
    main()
  File "/Users/alancoding/.virtualenvs/ansible29/lib/python3.7/site-packages/ansible_test/_data/sanity/validate-modules/validate_modules/main.py", line 1901, in main
    run()
  File "/Users/alancoding/.virtualenvs/ansible29/lib/python3.7/site-packages/ansible_test/_data/sanity/validate-modules/validate_modules/main.py", line 1799, in run
    mv1.validate()
  File "/Users/alancoding/.virtualenvs/ansible29/lib/python3.7/site-packages/ansible_test/_data/sanity/validate-modules/validate_modules/main.py", line 1621, in validate
    doc_info, docs = self._validate_docs()
  File "/Users/alancoding/.virtualenvs/ansible29/lib/python3.7/site-packages/ansible_test/_data/sanity/validate-modules/validate_modules/main.py", line 1013, in _validate_docs
    'invalid-documentation',
  File "/Users/alancoding/.virtualenvs/ansible29/lib/python3.7/site-packages/ansible_test/_data/sanity/validate-modules/validate_modules/main.py", line 851, in _validate_docs_schema
    for error in e.errors:
AttributeError: 'TypeError' object has no attribute 'errors'
```

The root cause appears to be due to `~/.ansible/collections/ansible_collections/community/aws//plugins/modules/redshift.py` and one other module (`plugins/modules/redshift_subnet_group.py`). See in specific:

```
DOCUMENTATION = '''
---
author:
version_added: 1.0.0
  - "Jens Carl (@j-carl), Hothead Games Inc."
  - "Rafael Driutti (@rafaeldriutti)"
module: redshift
short_description: create, delete, or modify an Amazon Redshift instance
```

Link:

https://github.com/ansible-collections/community.aws/blob/3b56fbdef774f26d9e2564507a0e686367224b11/plugins/modules/redshift.py#L13

The author value is none there.

I really don't care if users are allowed to do this or not, I just don't want it to prevent the tests from running. There's a valid point that CI is currently broken due to this.

This is modeled after @sivel's fix in https://github.com/ansible/ansible/pull/56882

### CODE REVIEW

**Review by sivel — APPROVED** [review]

**Comment by AlanCoding** [comment]
Link content side fix https://github.com/ansible-collections/community.aws/pull/128

File: test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/schema.py
Comment (felixfontein) [inline test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/schema.py]: I think it's better to have `if not isinstance(line, string_types): continue`. That way all crashes are prevented, and not just the special case when `author` is `None` or a list containing `None`.

The schema itself already checks whether `author` is either `None`, a string, or a list of strings, and will report everything that is none of these.

Comment (AlanCoding) [reply test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/schema.py]: I agree with checking against string types. That was my original direction.

But I don't think the schema check is working correctly without a patch of some sort here.

Comment (felixfontein) [reply test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/schema.py]: The schema check works correctly. (Which isn't necessarily the way you expect it to work :) ) The problem is the `author` function, which crashes for some inputs. It must not crash, and either return normally or raise a proper error like `Invalid`.

File: changelogs/fragments/70507-validate-null-author.yaml
Comment (felixfontein) [inline changelogs/fragments/70507-validate-null-author.yaml]:
```suggestion
  - Fixes ansible-test traceback when plugin author is not a string or a list of strings (https://github.com/ansible/ansible/pull/70507)
```

File: test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/schema.py
Comment (felixfontein) [inline test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/schema.py]:
```suggestion
            raise Invalid("Specify author as a string or a list of strings, or not specify author at all")
```

File: test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/schema.py
Comment (AlanCoding) [reply test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/schema.py]: Sorry, my eyes kind of glossed over the `continue` in your suggestion.

Testing that... it doesn't seem to work. In the updated PR now it gives

```
Running sanity test 'validate-modules' with Python 3.7
ERROR: Found 2 validate-modules issue(s) which need to be resolved:
ERROR: plugins/modules/redshift.py:0:0: invalid-documentation: DOCUMENTATION.author: Specify author as a list of strings for dictionary value @ data['author']. Got None
ERROR: plugins/modules/redshift_subnet_group.py:0:0: invalid-documentation: DOCUMENTATION.author: Specify author as a list of strings for dictionary value @ data['author']. Got None
See documentation for help: https://docs.ansible.com/ansible/devel/dev_guide/testing/sanity/validate-modules.html
```

But if I use the `continue` where the `raise` is now, there are no `validate-modules` failures, and no mention of the problematic `redshift.py` module.

I don't know why it's not giving the schema failure, in this case it's that `"author": None`, not `"author": [None]`, in the actual DOCUMENTATION string.

Comment (felixfontein) [reply test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/schema.py]: `None` is a valid value for `author` though, according to the schema. If the schema is correct, no error should be reported for the redshift module - or at least not for `author`.

Once `version_added` is checked for collections, it would have complained about that one, since `"1.0.0 - \"Jens Carl (@j-carl), Hothead Games Inc.\" - \"Rafael Driutti (@rafaeldriutti)\""` is NOT a valid version number.

Comment (AlanCoding) [reply test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/schema.py]: > None is a valid value for author though, according to the schema. If the schema is correct, no error should be reported for the redshift module - or at least not for author.

I don't wish to take a position on this. The only thing I'm sure about is that it shouldn't give that TypeError in this case which can only be debugged mucking with the Ansible source code itself. Whether the new behavior is pass or fail... I only had a slight leaning toward fail.

I don't know who the ideal referee is for `ansible-test` is, but if I did I would `@` them here so I can defer to their judgement.

Comment (felixfontein) [reply test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/schema.py]: I don't know either. Usually such questions are discussed in the public project meeting, if nobody from core team decides before that happens (looking at @sivel :) ).

In any case, if an empty author is not acceptable, you should better fix the schema and don't do this error reporting in the `author` function, or you change the schema to just run this function and not do other checks, and do better error reporting (right now there's only a generic `Specify author as a string or a list of strings, or not specify author at all` error which doesn't tell what it actually encountered, only that it didn't like the input).

Comment (sivel) [reply test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/schema.py]: Since the existing voluptuous schema defines `None` as a valid option, we should probably make the `author` function adhere to that as well.  However the schema is tricky.  `author: null` is allowed, however `author: [null]` is not.

I think we might need something like this:

```diff
diff --git a/test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/schema.py b/test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/schema.py
index 22b1a53d11..9587b29c98 100644
--- a/test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/schema.py
+++ b/test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/schema.py
@@ -412,15 +412,21 @@ def deprecation_schema(for_collection):
 
 
 def author(value):
+    if value is None:
+        return value
 
     if not is_iterable(value):
         value = [value]
 
     for line in value:
+        if not isinstance(line, string_types):
+            continue
         m = author_line.search(line)
         if not m:
             raise Invalid("Invalid author")
 
+    return value
+
 
 def doc_schema(module_name, for_collection=False, deprecated_module=False):
 ```

Comment (AlanCoding) [reply test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/schema.py]: I committed that, but it does change this case of redshift into a pass, not a failure.

Comment (felixfontein) [reply test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/schema.py]: The redshift module says `author: null` (where `null` is implicit), so it is correct with regard to satisfying the schema for `author`. Sooner or later we will have a check for `version_added` in collections, and that check would flag this case (since here `version_added` is garbled).

Comment (sivel) [reply test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/schema.py]: Yep, as of now this allows `author` to pass, since it actually does satisfy the schema.  The real issue in `community.aws` was the `version_added` was placed in the wrong spot, which actually invalidates `version_added`.

Unfortunately we don't validate `version_added` in collections currently, because at the time there had been no standard, and that is still being worked out to some extent.

Comment (felixfontein) [reply test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/schema.py]: That will come soonish, I hope. Whenever all more pressing things are done and I find time to work on my version_added PR again ;)

**Review by felixfontein — APPROVED** [review]

**Review by sivel — APPROVED** [review]

**Comment by felixfontein** [comment]
@AlanCoding do you want to create a backport to stable-2.10 for this?

File: test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/schema.py
Comment (felixfontein) [reply test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/schema.py]: This PR adds `version_added` validation: #70869. It was already there for module options and return values, but not for the top-level `version_added` of a module in a collection.

**Comment by felixfontein** [comment]
I've created a backport (#70873).

**Pull request merged**
